service/history/replication/raw_task_converter.go

```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
+		targetClusterID:  targetClusterID,
+		executionManager: executionManager,
+		mutableState:     mutableState,
+		logger:           logger,
+		progress:         progress,
+	}
+}
+
+func (c *syncVersionedTransitionTaskConverter) convert() (*replicationspb.ReplicationTask, error) {
+	// 1. task versioned transition not on current transition history
+	//   1.1 task has no events and no new run                   : no-op
+	//   1.2 task.VersionedTransition is found in cache 				 : verify [task]
+	//   1.3 task.VersionedTransition is not found in cache      : backfill events [task]
+	// 2. task versioned transition on current transition history
+	//   2.1 never synced                                        : SyncStateMutation: events [task, current]
+	//   2.2 task.VersionedTransition is found in cache          : verify [task]
+	//   2.3 last sync on current transition history             : SyncStateMutation: events (LCA(cache,current), current]
+	//   2.4 last sync not on current transition history         : SyncStateSnapshot: events (LCA(lastSync,current),current]
+
+	transitionHistory := c.mutableState.GetExecutionInfo().TransitionHistory
+	if !c.onCurrentBranch(c.taskInfo.VersionedTransition) {
+		// 1.1 task has no events and no new run
+		if c.taskInfo.FirstEventID == c.taskInfo.NextEventID && len(c.taskInfo.NewRunID) == 0 {
+			return nil, nil
+		}
+		if c.progress.TaskSent(c.taskInfo.VersionedTransition) {
+			return c.generateVerifyVersionedTransitionTask()
+		}
+		return c.generateBackfillHistoryTask()
+	}
+
+	lastSyncedTransition := c.progress.LastSyncedTransition()
+	if lastSyncedTransition == nil {
+		return c.generateSyncStateMutationFromTask(transitionHistory)
+	}
+	if c.progress.TaskSent(c.taskInfo.VersionedTransition) {
+		return c.generateVerifyVersionedTransitionTask()
+	}
+
+	lca, err := c.findLCAVersionHistoryItemFromCache()
+	if err != nil {
+		return nil, err
+	}
+	if c.onCurrentBranch(lastSyncedTransition) {
+		return c.generateSyncStateMutationFromCache(transitionHistory, lca)
+	}
+	return c.generateSyncStateSnapshotFromCache(transitionHistory, lca)
+}
+
+func (c *syncVersionedTransitionTaskConverter) onCurrentBranch(versionedTransition *persistencespb.VersionedTransition) bool {
+	if c.progress.LastSyncedTransition() == nil {
+		return false
+	}
+	return workflow.TransitionHistoryStalenessCheck(
+		c.mutableState.GetExecutionInfo().TransitionHistory,
+		versionedTransition,
+	) == nil
+}
+
+func (c *syncVersionedTransitionTaskConverter) getEvents(
+	transitionHistory []*persistencespb.VersionedTransition,
+	startEventID int64,
+	startEventVersion int64,
+	startEventInclusive bool,
+) ([]*commonpb.DataBlob, []byte, error) {
+	namespaceName, err := c.shardContext.GetNamespaceRegistry().GetNamespaceName(namespace.ID(c.taskInfo.NamespaceID))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resp, err := c.historyEngine.GetWorkflowExecutionRawHistoryV2(c.ctx, &historyservice.GetWorkflowExecutionRawHistoryV2Request{
```

Comment:
hmm does it work if we just make the persistence call to fetch raw events, like what we did in `getEventsBlob`. maybe we can even just reuse `getEventsBlob`? I could be missing something here though.

Reply:
will solve this in another PR combining with the SyncState API.